File: dotnet/samples/Varar.Sample/counter.steps.cs
using Varar;
using Varar.Core;

namespace Varar.Sample;

// A sample step file: a static Register(Registry) is the injected-Registrar entry point the
// adapter discovers by reflection; DefineState threads the initial state into the registry.
public static class CounterSteps
{
    public static Registry Register(Registry r)
    {
        var s = Steps.From(r);
        s.DefineState(() => Value.Map([new("count", Value.Of(0))]));
        s.Stimulus("I increment", state => Value.Map([new("count", Value.Of(state["count"].AsInt() + 1))]));
        s.Sensor("The count is {int}", (state, n) =>
        {
            var count = state["count"].AsInt();
            var expected = n.AsInt();
            if (count != expected)
            {
                throw new HandlerException($"expected {expected} but got {count}");
            }

            return null;
        });
        return s.ToRegistry();
    }
}

File: java/junit/src/test/java/dev/varar/junit/fixtures/CounterSteps.java
package dev.varar.junit.fixtures;

import dev.varar.State;
import dev.varar.StepDefinitions;
import dev.varar.Steps;

/**
 * A {@code state} fixture whose {@code action} step is ADDITIVE (unlike {@link
 * WidgetSteps}' absolute-set {@code context} step) — used by {@code
 * VarExampleDescriptorExecutionTest} to prove state does NOT leak between two examples in
 * the same file: if a second example's initial state were anything other than the fresh
 * {@code Ctx(0)} {@link Steps#defineState} produces, its counter would come out wrong,
 * not merely "still correct by coincidence" (the risk an absolute-set step like {@code
 * WidgetSteps}' can't expose).
 */
public final class CounterSteps implements StepDefinitions<CounterSteps.Ctx> {

    public record Ctx(int count) implements State {}

    @Override
    public void register(Steps<Ctx> s) {
        s.defineState(() -> new Ctx(0));
        s.stimulus("I add {int} to the counter", (Ctx ctx, Integer n) -> new Ctx(ctx.count() + n));
        s.sensor("the counter should be {int}", (Ctx ctx, Integer expected) -> ctx.count());
    }
}
